store/src/java-test/com/zimbra/cs/event/analytics/EventMetricTest.java

```diff
@@ -0,0 +1,270 @@
+        metric.increment(null);
+        metric.increment(null);
+        assertEquals("value after two increments should be 2", (Integer)2, metric.getValue());
+    }
+
+    @Test
+    public void testRatioMetric() throws Exception {
+
+        MetricInitializer<RatioMetric, Double, RatioIncrement> initializer = new MetricInitializer<RatioMetric, Double, RatioIncrement>() {
+
+            @Override
+            public RatioMetric getInitialData() {
+                return new RatioMetric(0d, 1);
+            }
+
+            @Override
+            public long getMetricLifetime() {
+                return 0;
+            }
+        };
+
+        EventMetric<RatioMetric, Double, RatioIncrement> metric = new EventMetric<RatioMetric, Double, RatioIncrement>("testAccountID", null, initializer) {
+
+            @Override
+            protected RatioIncrement getIncrement(List<Event> events) throws ServiceException {
+                return new RatioIncrement(1d, 2);
+            }
+        };
+
+        assertEquals("initial value should be 0", new Double(0), metric.getValue());
+        metric.increment(null);
+        assertEquals("new value should be 1/3", new Double(1d/3), metric.getValue());
+        metric.increment(null);
+        assertEquals("new value should be 2/5", new Double(2d/5), metric.getValue());
+    }
+
+    private void logContactFrequencyEvents(String contactEmail) {
+        logger.log(Event.generateEvent(ACCT_ID, 1, contactEmail, "me@zimbra.com", EventType.RECEIVED, null, null, System.currentTimeMillis()));
+        logger.log(Event.generateSentEvent(ACCT_ID, 2, "me@zimbra.com", contactEmail, null, null, System.currentTimeMillis()));
+        logger.log(Event.generateSentEvent(ACCT_ID, 3, "me@zimbra.com", "other@zimbra.com", null, null, System.currentTimeMillis()));
+        logger.log(Event.generateSentEvent("otherAccountId", 4, "me@zimbra.com", contactEmail, null, null, System.currentTimeMillis()));
+        logger.sendAllBatched();
+    }
+
+    private void logMsgEvents(String contactEmail, int startingMsgId, int numEvents, EventType eventType) {
+        logMsgEvents(contactEmail, startingMsgId, numEvents, eventType, System.currentTimeMillis());
+    }
+
+    private void logMsgEvents(String contactEmail, int startingMsgId, int numEvents, EventType eventType, long timestamp) {
+        for (int i=0; i<numEvents; i++) {
+            logger.log(Event.generateEvent(ACCT_ID, startingMsgId+i, contactEmail, "me@zimbra.com", eventType, null, null, timestamp));
+        }
+        //log an event for a different account id to make it doesn't affect the metrics
+        logger.log(Event.generateEvent("other", startingMsgId, contactEmail, "me@zimbra.com", eventType, null, null, timestamp));
+        logger.sendAllBatched();
+    }
+
+    @Test
+    public void testContactFrequency() throws Exception {
+
+        AtomicBoolean initialized = new AtomicBoolean(false);
```

Comment:
why do you need to use an AtomicBoolean and not simple boolean here and in similar places?

Reply:
I originally had dummy initializers defined as anonymous classes, with the getInitialData() calls setting a flag outside the class scope; I had to use an AtomicBoolean for this. When I refactored that to an inner class, I kept the AtomicBoolean even though it's unnecessary. I'll change this.